test/distributed/checkpoint/test_state_dict_stager.py

```diff
@@ -818,5 +827,424 @@ def test_dtensor(self):
+                "Shard offsets should match",
+            )
+            self.assertEqual(
+                local_shard.metadata.shard_sizes,
+                expected_metadata.shard_sizes,
+                "Shard sizes should match",
+            )
+
+        # Verify regular tensors
+        if "rank_scalar" in replicated_dict:
+            self.assertEqual(
+                replicated_dict["rank_scalar"].item(),
+                float(partner_rank),
+                f"Rank scalar should be {partner_rank}, got {replicated_dict['rank_scalar'].item()}",
+            )
+
+    @with_comms
+    def test_replication_basic(self):
+        """Test basic replication functionality with world_size=16"""
+        world_size = dist.get_world_size()
+
+        current_rank = dist.get_rank()
+
+        # Create unique DTensor state_dict for this rank
+        state_dict = self._create_simple_state_dict(current_rank)
+
+        # Initialize replication stager
+        stager = _ReplicationStager(
+            pg=dist.distributed_c10d._get_default_group(),
+            timeout=timedelta(seconds=30),
+            device=torch.device("cpu"),
+        )
+
+        # Perform replication
+        replicated_dict = stager.stage(state_dict)
+
+        # Calculate expected partner rank
+        partner_rank = (current_rank + world_size // 2) % world_size
+
+        # Verify DTensor replication
+        self._verify_simple_state_dict_replication(
+            replicated_dict, current_rank, partner_rank
+        )
+
+        # Clean up
+        stager.close()
+
+    @with_comms
+    def test_replication_dtensors(self):
+        """Test replication with DTensor and mixed tensor types"""
+        world_size = dist.get_world_size()
+
+        current_rank = dist.get_rank()
+
+        # Create CPU-based DeviceMesh for DTensor
+        device_mesh = DeviceMesh("cpu", list(range(world_size)))
+
+        # Create DTensor state_dict which includes different tensor types
+        state_dict = self._create_dtensor_state_dict(current_rank, device_mesh)
+
+        # Initialize replication stager
+        stager = _ReplicationStager(
+            pg=dist.distributed_c10d._get_default_group(),
+            timeout=timedelta(seconds=30),
+            device=torch.device("cpu"),
+        )
+
+        # Perform replication
+        result = stager.stage(state_dict)
+
+        # Wait for completion
+        from concurrent.futures import Future
+
+        if isinstance(result, Future):
+            replicated_dict = result.result()
+        else:
+            replicated_dict = result
+
+        # Calculate expected partner
+        partner_rank = (current_rank + world_size // 2) % world_size
+
+        # Verify all DTensor types are correctly replicated
+        self._verify_dtensor_replication(replicated_dict, current_rank, partner_rank)
+
+        # Clean up
+        stager.close()
+
+    @with_comms
+    def test_replication_sharded_tensors(self):
```

Comment:
Thanks for these test pointers. I'll try to get sharded tensor working. If you want to comment out these tests for now in order to land, feel free